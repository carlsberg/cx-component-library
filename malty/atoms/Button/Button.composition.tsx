import { MaltyThemeProvider } from '@carlsberg/malty.theme.malty-theme-provider';
import React from 'react';
import { Button } from './Button';
import { ButtonType } from './Button.types';

export const BasicButton = () => (
  <MaltyThemeProvider theme="global">
    <Button buttonType={ButtonType.Primary} text="Clickable button" />
  </MaltyThemeProvider>
);
