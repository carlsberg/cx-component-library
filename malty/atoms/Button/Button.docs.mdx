---
labels: ['react', 'typescript', 'button']
description: 'A Button component.'
---

import { Button, ButtonType } from '.';
import { MaltyThemeProvider } from '@carlsberg/malty.theme.malty-theme-provider';

Button example:

A simple Button component with some text

// live component playground - if you remove the word live it turns into a code snippet
```js live
  <MaltyThemeProvider theme="global">
    <Button buttonType={ButtonType.Primary} text="Clickable button" />
  </MaltyThemeProvider>
```
