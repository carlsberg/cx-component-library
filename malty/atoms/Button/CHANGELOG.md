# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- **button:** Inherit font properties from button definition ([#327](https://github.com/CarlsbergGBS/cx-component-library/issues/327)) ([b8a59bf](https://github.com/CarlsbergGBS/cx-component-library/commit/b8a59bf5d5498b48e75288081af94c4b7a9c24da))
- **DSM-245:** improvements on button ([#202](https://github.com/CarlsbergGBS/cx-component-library/issues/202)) ([774968b](https://github.com/CarlsbergGBS/cx-component-library/commit/774968bb41a1275184c65ea7ec9859188a1463a0))
- **DSM-333:** button loading ([#351](https://github.com/CarlsbergGBS/cx-component-library/issues/351)) ([2dad7d9](https://github.com/CarlsbergGBS/cx-component-library/commit/2dad7d929132f773318ffb9002f34889cac5e372))
- **DSM-333:** fullWidth button ([#313](https://github.com/CarlsbergGBS/cx-component-library/issues/313)) ([efd088d](https://github.com/CarlsbergGBS/cx-component-library/commit/efd088d66f7c49d45a73866afe1ddf24fb2bf505))
- Remove typography provider from every component ([#383](https://github.com/CarlsbergGBS/cx-component-library/issues/383)) ([5d05680](https://github.com/CarlsbergGBS/cx-component-library/commit/5d05680e40d3da5bb3042bd8191427a70606afc3))
- reverted changes ([#396](https://github.com/CarlsbergGBS/cx-component-library/issues/396)) ([ecfcf57](https://github.com/CarlsbergGBS/cx-component-library/commit/ecfcf573863253fea16e24764b6f452c3cc4f215))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- **button, checkbox:** [DSM-695] Add tests ([#685](https://github.com/CarlsbergGBS/cx-component-library/issues/685)) ([7226143](https://github.com/CarlsbergGBS/cx-component-library/commit/72261439f6c841f7b4e9c6c42ab7f13d4cbc0b41))
- **button, chip:** Remove haveStyle checks ([#708](https://github.com/CarlsbergGBS/cx-component-library/issues/708)) ([5c9762d](https://github.com/CarlsbergGBS/cx-component-library/commit/5c9762d63041ad967e7b8390d6fa32c59671a64c))
- **button:** [DSM-610] Square button with only icon ([#697](https://github.com/CarlsbergGBS/cx-component-library/issues/697)) ([040d317](https://github.com/CarlsbergGBS/cx-component-library/commit/040d317a75e0c48e979b8db1d2b41457c1444c19))
- **button:** [DSM-696] update hover and active behaviours in Primary variation ([#737](https://github.com/CarlsbergGBS/cx-component-library/issues/737)) ([7002d7e](https://github.com/CarlsbergGBS/cx-component-library/commit/7002d7e1fe2292606a176807efa657cd610f43b5))
- **button:** [DSM-845] add selected styles ([#835](https://github.com/CarlsbergGBS/cx-component-library/issues/835)) ([ff76591](https://github.com/CarlsbergGBS/cx-component-library/commit/ff76591d6beee24ca15fe9e857bcd519cf592876))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **chip:** [DSM-700] Add tests ([#695](https://github.com/CarlsbergGBS/cx-component-library/issues/695)) ([45ddade](https://github.com/CarlsbergGBS/cx-component-library/commit/45ddadea01a0c56597032694df969fc5f2058559))
- **DSM-198:** Chips component ([#229](https://github.com/CarlsbergGBS/cx-component-library/issues/229)) ([c31055c](https://github.com/CarlsbergGBS/cx-component-library/commit/c31055c9adeab246b63af5af627d83bf04cbd60a))
- **DSM-228:** Added rating component ([#379](https://github.com/CarlsbergGBS/cx-component-library/issues/379)) ([ab47354](https://github.com/CarlsbergGBS/cx-component-library/commit/ab47354db803fb051a1f87d66ee34966698df8bd))
- **DSM-229:** button ([#165](https://github.com/CarlsbergGBS/cx-component-library/issues/165)) ([1950717](https://github.com/CarlsbergGBS/cx-component-library/commit/19507172b731f341ecf32c13eac07e465c6aa659))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-256:** link ([#279](https://github.com/CarlsbergGBS/cx-component-library/issues/279)) ([7185428](https://github.com/CarlsbergGBS/cx-component-library/commit/718542872ecffe4a580bfeb8bf71712317f8bad4))
- **DSM-337:** button color ([#295](https://github.com/CarlsbergGBS/cx-component-library/issues/295)) ([31349ca](https://github.com/CarlsbergGBS/cx-component-library/commit/31349cadbbef7d0d5915b150349a3f7c90fc01ff))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** Remove `react/jsx-props-no-spreading` comments ([#558](https://github.com/CarlsbergGBS/cx-component-library/issues/558)) ([823444c](https://github.com/CarlsbergGBS/cx-component-library/commit/823444ce0800bc7f49b584de59d247eb8497f559))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- **loading-overlay:** create component ([#856](https://github.com/CarlsbergGBS/cx-component-library/issues/856)) ([32e0597](https://github.com/CarlsbergGBS/cx-component-library/commit/32e05974e0240b4d317183900469b2abf0a957e3))
- New Chip component and minor changes ([#239](https://github.com/CarlsbergGBS/cx-component-library/issues/239)) ([ce92ab0](https://github.com/CarlsbergGBS/cx-component-library/commit/ce92ab0543b96296952a29e2d43d0df29d4e188f))
- new icon colors ([#215](https://github.com/CarlsbergGBS/cx-component-library/issues/215)) ([9247d33](https://github.com/CarlsbergGBS/cx-component-library/commit/9247d3370bcde78432033ba8b23837f97e0b0dc0))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
- Version bumps for various fixes ([#205](https://github.com/CarlsbergGBS/cx-component-library/issues/205)) ([57c2e30](https://github.com/CarlsbergGBS/cx-component-library/commit/57c2e30804ba0aa6ebdd6a3ac59044b69ad9393f))
