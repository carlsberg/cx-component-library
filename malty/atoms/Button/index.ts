export { Button } from './Button';
export { ButtonStyle, ButtonTypes, IconPosition, SizeTypes } from './Button.types';
export type { ButtonProps } from './Button.types';
