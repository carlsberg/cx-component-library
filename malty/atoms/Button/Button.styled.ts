import styled, { css, keyframes } from 'styled-components';
import { ButtonIconPosition } from './Button.types';

const animateShow = keyframes`
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
`;

export const StyledAnchor = styled.a`
  text-decoration: none;
`;

const StyledButton = styled.button<{
  hasText: boolean;
  hasIcon: boolean;
  loading: boolean;
  isNegative?: boolean;
  fullWidth?: boolean;
  sizing: string;
  horizontalPadding: string;
  fontSize: string;
  iconSize: string;
  iconPos: ButtonIconPosition;
}>`
  font-family: inherit;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  padding: ${({ theme }) => `0 ${theme.sizes.s.value}`};
  height: ${({ sizing }) => `${sizing}`};
  font-size: ${({ fontSize }) => `${fontSize}`};
  font-weight: bold;
  transition: 0.25s ease-in-out;
  transition-property: background-color, color;
  cursor: pointer;
  border: none;
  gap: ${({ theme }) => theme.sizes.s.value};
  width: ${({ fullWidth }) => (fullWidth ? '100%' : 'auto')};
  flex-direction: ${({ iconPos }) =>
    ButtonIconPosition[iconPos] === ButtonIconPosition.Right ? 'row' : 'row-reverse'};
  &:hover,
  &:focus {
    outline: 0;
  }

  &:disabled {
    cursor: default;
    color: ${({ theme }) => theme.colors.colours.default.white.value};
    background-color: ${({ theme }) => theme.colors.colours.system['disable-light-theme'].value};
    &:hover {
      color: ${({ theme }) => theme.colors.colours.default.white.value};
      background-color: ${({ theme }) => theme.colors.colours.system['disable-light-theme'].value};
    }
  }

  .text-container {
    display: flex;
    align-items: center;
    gap: ${({ theme }) => theme.sizes['2xs'].value};
    opacity: 1;
    &.invisible {
      opacity: 0;
    }
  }

  .secondary-container {
    display: flex;
    justify-content: center;
    align-items: center;
    animation: ${animateShow} 0.25s linear;
    > div {
      gap: 0;
      padding: 0;
    }
  }

  svg {
    height: ${({ iconSize }) => `${iconSize}`};
    width: ${({ iconSize }) => `${iconSize}`};
  }

  ${({ hasText, hasIcon, sizing, loading }) =>
    ((!hasText && hasIcon) || loading) &&
    css`
      padding: 0;
      justify-content: center;
      width: ${sizing};
    `};
`;

export const StyledPrimaryButton = styled(StyledButton)`
  background-color: ${({ isNegative, theme }) =>
    isNegative ? theme.colors.colours.default.white.value : theme.colors.colours.default['digital-black'].value};
  color: ${({ isNegative, theme }) =>
    isNegative ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.default.white.value};
  border: 1px solid ${({ theme }) => theme.colors.colours.default.transparent.value};

  &:hover {
    opacity: 0.95;
  }
  &:active {
    opacity: 0.9;
  }
  &:disabled {
    ${({ isNegative, theme }) =>
      isNegative &&
      css`
        background-color: ${theme.colors.colours.system['disable-dark-theme'].value};
        color: ${theme.colors.colours.system['disable-dark-theme'].value};
        &:hover {
          background-color: ${theme.colors.colours.system['disable-dark-theme'].value};
          color: ${theme.colors.colours.system['disable-dark-theme'].value};
        }
        svg {
          fill: ${theme.colors.colours.system['disable-dark-theme'].value};
          color: ${theme.colors.colours.system['disable-dark-theme'].value};
        }
      `};
  }
`;

export const StyledSecondaryButton = styled(StyledButton)`
  background-color: ${({ theme }) => theme.colors.colours.default.transparent.value};
  color: ${({ isNegative, theme }) =>
    isNegative ? theme.colors.colours.default.white.value : theme.colors.colours.default['digital-black'].value};
  border: 1px solid
    ${({ isNegative, theme }) =>
      isNegative ? theme.colors.colours.default.white.value : theme.colors.colours.default['digital-black'].value};
  &:hover {
    background-color: ${({ isNegative, theme }) =>
      isNegative
        ? theme.colors.colours.overlay.white[5].value
        : theme.colors.colours.overlay['digital-black'][5].value};
  }
  &:active {
    background-color: ${({ isNegative, theme }) =>
      isNegative
        ? theme.colors.colours.overlay.white[10].value
        : theme.colors.colours.overlay['digital-black'][10].value};
  }
  &:disabled {
    svg {
      fill: ${({ isNegative, theme }) =>
        isNegative
          ? theme.colors.colours.system['disable-dark-theme'].value
          : theme.colors.colours.system['disable-light-theme'].value};
      color: ${({ isNegative, theme }) =>
        isNegative
          ? theme.colors.colours.system['disable-dark-theme'].value
          : theme.colors.colours.system['disable-light-theme'].value};
    }
    background-color: ${({ theme }) => theme.colors.colours.default.transparent.value};
    border: 1px solid
      ${({ isNegative, theme }) =>
        isNegative
          ? theme.colors.colours.system['disable-dark-theme'].value
          : theme.colors.colours.system['disable-light-theme'].value};
    color: ${({ isNegative, theme }) =>
      isNegative
        ? theme.colors.colours.system['disable-dark-theme'].value
        : theme.colors.colours.system['disable-light-theme'].value};
    &:hover {
      background-color: ${({ theme }) => theme.colors.colours.default.transparent.value};
      border: 1px solid
        ${({ isNegative, theme }) =>
          isNegative
            ? theme.colors.colours.system['disable-dark-theme'].value
            : theme.colors.colours.system['disable-light-theme'].value};
      color: ${({ isNegative, theme }) =>
        isNegative
          ? theme.colors.colours.system['disable-dark-theme'].value
          : theme.colors.colours.system['disable-light-theme'].value};
    }
  }
`;

export const StyledTransparentButton = styled(StyledButton)`
  background-color: ${({ theme }) => theme.colors.colours.default.transparent.value};

  color: ${({ isNegative, theme }) =>
    isNegative ? theme.colors.colours.default.white.value : theme.colors.colours.default['digital-black'].value};
  &.active {
    background-color: ${({ theme }) => theme.colors.colours.support[40].value};
  }
  &:hover {
    background-color: ${({ isNegative, theme }) =>
      isNegative
        ? theme.colors.colours.overlay.white[5].value
        : theme.colors.colours.overlay['digital-black'][5].value};
  }
  &:active {
    background-color: ${({ isNegative, theme }) =>
      isNegative
        ? theme.colors.colours.overlay.white[10].value
        : theme.colors.colours.overlay['digital-black'][10].value};
  }
  &:disabled {
    svg {
      fill: ${({ isNegative, theme }) =>
        isNegative
          ? theme.colors.colours.system['disable-dark-theme'].value
          : theme.colors.colours.system['disable-light-theme'].value};
      color: ${({ isNegative, theme }) =>
        isNegative
          ? theme.colors.colours.system['disable-dark-theme'].value
          : theme.colors.colours.system['disable-light-theme'].value};
    }
    background-color: ${({ theme }) => theme.colors.colours.default.transparent.value};
    color: ${({ isNegative, theme }) =>
      isNegative
        ? theme.colors.colours.system['disable-dark-theme'].value
        : theme.colors.colours.system['disable-light-theme'].value};
    &:hover {
      background-color: ${({ theme }) => theme.colors.colours.default.transparent.value};
      color: ${({ isNegative, theme }) =>
        isNegative
          ? theme.colors.colours.system['disable-dark-theme'].value
          : theme.colors.colours.system['disable-light-theme'].value};
    }
  }
`;

export const StyledLinkButton = styled(StyledButton)`
  background-color: transparent;
  color: ${({ isNegative, theme }) =>
    isNegative ? theme.colors.colours.default.white.value : theme.colors.colours.default['digital-black'].value};
  text-decoration: underline;
  padding: 0;
  height: auto;
  &:hover {
    background-color: ${({ theme }) => theme.colors.colours.default.transparent.value};
  }
`;
