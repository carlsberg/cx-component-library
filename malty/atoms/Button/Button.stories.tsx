import { IconNamesTypes } from '@carlsberggroup/malty.atoms.icon';
import { Story } from '@storybook/react';
import React from 'react';
import { ButtonProps, SizeTypes } from '.';
import { Button } from './Button';
import { ButtonStyle, ButtonTypes, IconPosition } from './Button.types';

export default {
  title: 'Atoms/Button',
  component: Button,
  argTypes: {
    text: { control: 'text' },
    type: {
      options: Object.values(ButtonTypes),
      control: {
        type: 'radio'
      },
      table: {
        defaultValue: {
          summary: 'submit'
        }
      }
    },
    style: {
      options: Object.values(ButtonStyle),
      control: {
        type: 'radio'
      }
    },
    scroll: {
      description: 'Scroll position where will floater show',
      defaultValue: 0,
      control: {
        type: 'number'
      }
    },
    loading: {
      table: {
        defaultValue: {
          summary: 'false'
        }
      },
      control: 'boolean'
    },
    success: { control: 'boolean' },
    successIcon: {
      options: Object.values(IconNamesTypes),
      description: 'Name options listed below',
      defaultValue: '',
      control: {
        type: 'select'
      }
    },
    successText: { control: 'text' },
    error: { control: 'boolean' },
    errorIcon: {
      options: Object.values(IconNamesTypes),
      description: 'Name options listed below',
      defaultValue: '',
      control: {
        type: 'select'
      }
    },
    icon: {
      options: Object.values(IconNamesTypes),
      description: 'Name options listed below',
      defaultValue: '',
      control: {
        type: 'select'
      }
    },
    size: {
      options: Object.values(SizeTypes),
      description: 'Size options are',
      table: {
        defaultValue: {
          summary: 'Medium'
        }
      },
      control: {
        type: 'radio'
      }
    },
    iconPos: {
      options: Object.values(IconPosition),
      description: 'Icon position',
      table: {
        defaultValue: {
          summary: 'Right'
        }
      },
      control: {
        type: 'radio'
      }
    },
    isWhite: { control: 'boolean' },
    disabled: {
      table: {
        defaultValue: {
          summary: 'false'
        }
      },
      control: 'boolean'
    },
    fullWidth: { control: 'boolean' },
    url: { control: 'text' },
    selected: {
      description: 'Add classname of "active" to element',
      control: 'boolean'
    },
    errorText: { control: 'text' },
    children: { control: 'text' }
  }
};

const Template: Story<ButtonProps> = (args) => <Button {...args} />;

export const Primary = Template.bind({});
Primary.args = {
  style: ButtonStyle.Primary,
  text: 'Primary',
  type: ButtonTypes.Submit,
  size: SizeTypes.Medium,
  iconPos: IconPosition.Right,
  loading: false,
  error: false,
  success: false,
  isWhite: false,
  disabled: false,
  fullWidth: false,
  url: '',
  selected: false,
  successIcon: IconNamesTypes.ItemCheck,
  successText: 'Success',
  errorIcon: IconNamesTypes.ItemClose,
  errorText: 'Error'
};

export const Secondary = Template.bind({});
Secondary.args = {
  style: ButtonStyle.Secondary,
  text: 'Secondary',
  type: ButtonTypes.Submit,
  size: SizeTypes.Medium,
  iconPos: IconPosition.Right,
  loading: false,
  error: false,
  success: false,
  isWhite: false,
  disabled: false,
  fullWidth: false,
  url: '',
  selected: false,
  successIcon: IconNamesTypes.ItemCheck,
  successText: 'Success',
  errorIcon: IconNamesTypes.ItemClose,
  errorText: 'Error'
};

export const Floater = Template.bind({});
Floater.args = {
  style: ButtonStyle.Floater,
  icon: IconNamesTypes.ArrowSmallUp,
  type: ButtonTypes.Submit,
  size: SizeTypes.Medium,
  iconPos: IconPosition.Right,
  loading: false,
  error: false,
  success: false,
  isWhite: false,
  disabled: false,
  fullWidth: false,
  url: '',
  selected: false,
  successIcon: IconNamesTypes.ItemCheck,
  successText: 'Success',
  errorIcon: IconNamesTypes.ItemClose,
  errorText: 'Error',
  scroll: 0
};

export const Link = Template.bind({});
Link.args = {
  style: ButtonStyle.Link,
  text: 'Link text',
  size: SizeTypes.Medium,
  iconPos: IconPosition.Right,
  loading: false,
  error: false,
  success: false,
  isWhite: false,
  disabled: false,
  fullWidth: false,
  url: '',
  selected: false,
  successIcon: IconNamesTypes.ItemCheck,
  successText: '',
  errorIcon: IconNamesTypes.ItemClose,
  errorText: ''
};
