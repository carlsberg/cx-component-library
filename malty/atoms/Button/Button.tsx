import { Icon, IconColors, IconNamesTypes, IconSizesTypes } from '@carlsberggroup/malty.atoms.icon';
import { Loading, LoadingStatus } from '@carlsberggroup/malty.molecules.loading';
import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { ButtonProps, Sizes, SizeTypes } from '.';
import {
  StyledAnchor,
  StyledFloaterButton,
  StyledLinkButton,
  StyledPrimaryButton,
  StyledSecondaryButton
} from './Button.styled';
import { ButtonStyle, ButtonTypes, IconPosition } from './Button.types';

export const Button = ({
  text,
  style,
  type = ButtonTypes.Submit,
  isWhite = false,
  fullWidth = false,
  selected = false,
  disabled,
  onClick,
  icon,
  url,
  size = SizeTypes.Medium,
  iconPos = IconPosition.Right,
  loading,
  error,
  errorIcon,
  errorText,
  success,
  successIcon,
  successText,
  children
}: ButtonProps) => {
  let Component = StyledPrimaryButton;
  let iconColor = isWhite ? IconColors.Primary : IconColors.White;
  switch (style) {
    case ButtonStyle.Secondary:
      Component = StyledSecondaryButton;
      iconColor = isWhite ? IconColors.White : IconColors.Primary;
      break;
    case ButtonStyle.Floater:
      Component = StyledFloaterButton;
      break;
    case ButtonStyle.Link:
      Component = StyledLinkButton;
      iconColor = isWhite ? IconColors.White : IconColors.Primary;
      break;
    default:
      break;
  }
  const theme = useContext(ThemeContext) || defaultTheme;
  const [numSize, setNumSize] = useState(Sizes.Medium);

  const renderComponent = () => (
    <Component
      type={type}
      disabled={disabled}
      hasText={!!text || !!children}
      hasIcon={!!icon}
      sizing={numSize}
      onClick={onClick}
      isWhite={isWhite}
      fullWidth={fullWidth}
      iconPos={iconPos}
      className={selected ? 'active' : ''}
      theme={theme}
    >
      {(success && successText) || (error && errorText) || text || children}
      {icon && !loading && <Icon name={icon} color={iconColor} size={IconSizesTypes.Small} />}
      {/* This is temporary use LoadingProps['status'] since the updated loading is not exported yet */}
      {loading && !success && !error && <Loading status={'Pending' as LoadingStatus} />}
      {success && !error && (
        <Icon name={successIcon || IconNamesTypes.ItemCheck} color={iconColor} size={IconSizesTypes.Small} />
      )}
      {!success && error && (
        <Icon name={errorIcon || IconNamesTypes.ItemClose} color={iconColor} size={IconSizesTypes.Small} />
      )}
    </Component>
  );

  useEffect(() => {
    switch (size) {
      case SizeTypes.Small: {
        setNumSize(Sizes.Small);
        break;
      }
      case SizeTypes.Large: {
        setNumSize(Sizes.Large);
        break;
      }
      case SizeTypes.XLarge: {
        setNumSize(Sizes.XLarge);
        break;
      }
      default: {
        setNumSize(Sizes.Medium);
        break;
      }
    }
  }, [size]);

  return url ? (
    <StyledAnchor target="_blank" href={url} rel="noreferrer" className={selected ? 'active' : ''}>
      {renderComponent()}
    </StyledAnchor>
  ) : (
    renderComponent()
  );
};
