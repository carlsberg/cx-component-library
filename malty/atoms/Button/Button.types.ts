import { IconNamesTypes } from '@carlsberggroup/malty.atoms.icon';
import React from 'react';

export interface ButtonProps {
  text?: string;
  icon?: IconNamesTypes;
  iconPos?: IconPosition;
  url?: string;
  buttonType: ButtonType;
  size?: SizeTypes;
  isWhite?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>;
  selected?: boolean;
  disabled?: boolean;
  loading?: boolean;
  error?: boolean;
  success?: boolean;
  fullWidth?: boolean;
  successIcon?: IconNamesTypes;
  successText?: string;
  errorIcon?: IconNamesTypes;
  errorText?: string;
}

export enum SizeTypes {
  Small = 'Small',
  Medium = 'Medium',
  Large = 'Large',
  XLarge = 'XLarge'
}

export enum Sizes {
  Small = 32,
  Medium = 40,
  Large = 48,
  XLarge = 56
}

export enum ButtonType {
  Primary = 'primary',
  Secondary = 'secondary',
  Floater = 'floater',
  Link = 'link'
}

export enum IconPosition {
  Right = 'Right',
  Left = 'Left'
}
