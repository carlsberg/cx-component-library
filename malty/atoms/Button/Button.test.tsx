import { jsonRenderer, render } from '@carlsberggroup/malty.utils.test';
import { fireEvent, screen } from '@testing-library/react';
import React from 'react';
import { Button } from './Button';
import { ButtonStyle } from './Button.types';

const defaultText = 'Submit';
const newText = 'Go';

describe('button', () => {
  it('matches snapshot', () => {
    const onClick = jest.fn();
    const view = jsonRenderer(<Button text={defaultText} style={ButtonStyle.Primary} onClick={onClick} disabled />);
    expect(view).toMatchSnapshot();
  });

  it('renders with correct text', () => {
    const { rerender } = render(<Button text={defaultText} style={ButtonStyle.Primary} />);
    expect(screen.getByText(defaultText)).not.toBeNull();

    rerender(<Button text={newText} style={ButtonStyle.Primary} />);
    expect(screen.getByText(newText)).not.toBeNull();
  });

  it('renders with correct text via child', () => {
    render(<Button style={ButtonStyle.Primary}>{defaultText}</Button>);
    expect(screen.getByText(defaultText)).not.toBeNull();
  });

  it('calls function on click', () => {
    const onClick = jest.fn();
    render(<Button text={defaultText} style={ButtonStyle.Primary} onClick={onClick} />);
    fireEvent.click(screen.getByText(defaultText));
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  it('does not call function on click when disabled', () => {
    const onClick = jest.fn();
    render(<Button text={defaultText} style={ButtonStyle.Primary} onClick={onClick} disabled />);
    fireEvent.click(screen.getByText(defaultText));
    expect(onClick).toHaveBeenCalledTimes(0);
  });
});
