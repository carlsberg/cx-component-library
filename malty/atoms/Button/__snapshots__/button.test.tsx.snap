// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button matches snapshot 1`] = `
<span
  className="sc-bczRLJ hROJNc"
>
  <button
    className="sc-gKXOVf sc-iBkjds foJVnI bnLSHM"
    disabled={true}
    fontSize="16px"
    onClick={[MockFunction]}
    tabIndex={-1}
    type="button"
  >
    <div
      className="text-container "
    >
      Submit
    </div>
  </button>
</span>
`;
